compiler/src/dotty/tools/dotc/ast/MainProxies.scala

```diff
@@ -114,4 +124,298 @@ object MainProxies {
+          val argName = nme.args ++ idx.toString
+          val isRepeated = formal.isRepeatedParam
+          val formalType = if isRepeated then formal.argTypes.head else formal
+          val getterName = if isRepeated then nme.varargGetter else nme.argGetter
+          val defaultValueGetterOpt = defaultValueSymbols.get(idx) match
+            case None => ref(defn.NoneModule.termRef)
+            case Some(dvSym) =>
+               val value = unitToValue(ref(dvSym.termRef))
+               Apply(ref(defn.SomeClass.companionModule.termRef), value)
+          val argGetter0 = TypeApply(Select(Ident(nme.cmd), getterName), TypeTree(formalType) :: Nil)
+          val argGetter =
+            if isRepeated then argGetter0
+            else Apply(argGetter0, List(Literal(Constant(idx)), defaultValueGetterOpt))
+
+          ValDef(argName, TypeTree(), argGetter)
+    end argValDefs
+
+
+    /** Create a list of argument references that will be passed as argument to the main method.
+     *  `args0`, ...`argn*`
+     */
+    def argRefs(mt: MethodType): List[Tree] =
+      for ((formal, paramName), idx) <- mt.paramInfos.zip(mt.paramNames).zipWithIndex yield
+        val argRef = Apply(Ident(nme.args ++ idx.toString), Nil)
+        if formal.isRepeatedParam then repeated(argRef) else argRef
+    end argRefs
+
+
+    /** Turns an annotation (e.g. `@main(40)`) into an instance of the class (e.g. `new scala.main(40)`). */
+    def instantiateAnnotation(annot: Annotation): Tree =
+      val argss = {
+        def recurse(t: tpd.Tree, acc: List[List[Tree]]): List[List[Tree]] = t match {
+          case Apply(t, args: List[tpd.Tree]) => recurse(t, extractArgs(args) :: acc)
+          case _ => acc
+        }
+
+        def extractArgs(args: List[tpd.Tree]): List[Tree] =
+          args.flatMap {
+            case Typed(SeqLiteral(varargs, _), _) => varargs.map(arg => TypedSplice(arg))
+            case arg: Select if arg.name.is(DefaultGetterName) => Nil  // Ignore default values, they will be added later by the compiler
+            case arg => List(TypedSplice(arg))
+          }
+
+        recurse(annot.tree, Nil)
+      }
+
+      New(TypeTree(annot.symbol.typeRef), argss)
+    end instantiateAnnotation
+
+    def generateMainClass(mainCall: Tree, args: List[Tree], parameterInfos: List[Tree]): TypeDef =
+      val cmdInfo =
+        val nameTree = Literal(Constant(mainFun.showName))
+        val docTree = Literal(Constant(documentation.mainDoc))
+        val paramInfos = Apply(ref(defn.SeqModule.termRef), parameterInfos)
+        New(TypeTree(defn.MainAnnotationCommandInfo.typeRef), List(List(nameTree, docTree, paramInfos)))
+
+      val cmd = ValDef(
+        nme.cmd,
+        TypeTree(),
+        Apply(
+          Select(instantiateAnnotation(mainAnnot), nme.command),
+          List(cmdInfo, Ident(nme.args))
+        )
+      )
+      val run = Apply(Select(Ident(nme.cmd), nme.run), mainCall)
+      val body = Block(cmdInfo :: cmd :: args, run)
```

Comment:
There is a `try` clause for the `@main` annotations but not for the custom `MainAnnotations` ones. The `try` block for the `@main` annotations checks if the parsing of arguments succeeded. This might be the right place to implement the flow control logic that would allow the user to terminate the program early, without using tricks like `System.exit(0)` or asserts. Maybe it is also a good idea to define what exceptions the custom annotation can throw as part of `Command`, and how `run` should handle them?

Reply:
At this point, we do not know what can or should be caught. Each implementation may have different parsers. 
Users can implement this however they want in the run. They can add the `try` and the proper cases or let it crash.

`System.exit(0)` is just an example. In that situation, we just wanted to terminate after taking an exceptional branch that does not require the arguments to be parsed.